File: mentoring/grpc/bank/account/bank_account.go
// Package account provides a bank account.
package account

import "sync"

// Account provides a bank account.
// If any Account method is called on an closed account, it must not modify
// the account and must return ok = false.
type Account struct {
	sync.RWMutex
	balance int64
	open    bool
}

// Open returns a new Account.
// If Open is given a negative initial deposit, it must return nil.
func Open(initialDeposit int64) *Account {
	if initialDeposit < 0 {
		return nil
	}
	return &Account{
		balance: initialDeposit,
		open:    true,
	}
}

// Close will close an account.
func (a *Account) Close() (payout int64, ok bool) {
	a.Lock()
	defer a.Unlock()
	if !a.open {
		return 0, false
	}
	a.open = false
	return a.balance, true
}

// Balance returns the account balance.
func (a *Account) Balance() (balance int64, ok bool) {
	a.RLock()
	defer a.RUnlock()
	if !a.open {
		return 0, false
	}
	return a.balance, true
}

// Deposit alters the account balance.
// It is thread safe.
// Deposit must handle a negative amount as a withdrawal. Withdrawals must
// not succeed if they result in a negative balance.
func (a *Account) Deposit(amount int64) (newBalance int64, ok bool) {
	a.Lock()
	defer a.Unlock()
	if !a.open || a.balance+amount < 0 {
		return 0, false
	}
	a.balance += amount
	return a.balance, true
}

File: python/bank-account/bank_account.py
import threading

class BankAccount(object):
  def __init__(self):
    self.lock = threading.Lock()
    self.balance = None

  def get_balance(self):
    with self.lock:
      if self.balance is None:
        raise ValueError('account not open')
      return self.balance

  def open(self):
    with self.lock:
      if self.balance is not None:
        raise ValueError('account already open')
      self.balance = 0

  def deposit(self, amount):
    if amount <= 0:
      raise ValueError('amount must be greater than 0')
    with self.lock:
      if self.balance is None:
        raise ValueError('account not open')
      self.balance += amount

  def withdraw(self, amount):
    if amount <= 0:
      raise ValueError('amount must be greater than 0')
    with self.lock:
      if self.balance is None:
        raise ValueError('account not open')
      if self.balance < amount:
        raise ValueError('amount must be less than balance')
      self.balance -= amount

  def close(self):
    with self.lock:
      if self.balance is None:
        raise ValueError('account not open')
      self.balance = None

# vim:ts=2:sw=2:expandtab
